src/Hint/NoCapitalisms.hs

```diff
@@ -0,0 +1,119 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE LambdaCase #-}
+{-
+    Detect uses of capitalisms
+
+    Only allow up to two consecutive capital letters in top level
+    identifiers.
+
+    Identifiers containing underscores are exempted from thus rule.
+    Identifiers of FFI bindings are exempted from thus rule.
+
+    Locally bound identifiers and module names are not checked.
+
+<TEST>
+data LHsDecl
+class FOO a where -- @Ignore
+class Foo a where getFOO :: Bool
+data Foo = Bar | BAAZ -- @Ignore
+data Foo = B_ar | BAAZ -- @Ignore
+data Foo = Bar | B_AAZ
+data OTPToken = OTPToken -- @Ignore
+data OTP_Token = Foo 
+sendSMS = _ -- @Ignore
+runTLS = _ -- @Ignore
+runTLSSocket = _ -- @Ignore
+runTLS_Socket
+newtype TLSSettings = TLSSettings -- @Ignore
+tlsSettings
+data CertSettings = CertSettings
+tlsServerHooks
+tlsServerDHEParams = _  -- @Ignore
+type WarpTLSException = () -- @Ignore
+get_SMS
+runCI
+foreign import ccall _FIREMISSLES :: IO ()
+getSMS :: IO () -- @Ignore
+gFOO = _ -- @Ignore
+geFOO = _ -- @Ignore
+getFOO = _ -- @Ignore
+</TEST>
+-}
+
+module Hint.NoCapitalisms(noCapitalismsHint) where
+
+import Hint.Type
+import Data.List.Extra as E
+import Data.List.NonEmpty as NE
+import Data.Char
+import Data.Maybe
+
+import GHC.Types.Basic
+import GHC.Types.SourceText
+import GHC.Data.FastString
+import GHC.Hs.Decls
+import GHC.Hs.Extension
+import GHC.Hs
+import GHC.Types.SrcLoc
+
+import Language.Haskell.GhclibParserEx.GHC.Hs.Decls
+import Language.Haskell.GhclibParserEx.GHC.Utils.Outputable
+import GHC.Util
+
+noCapitalismsHint :: DeclHint
+noCapitalismsHint _ _ decl = [ remark Ignore "Avoid capitalisms" (reLoc (shorten decl))
+                             | not $ isForD decl
+                             , name <- nubOrd $ getNames decl
+                             , not $ hasUnderscore name
+                             , hasCapitalism name
+                             ]
+
+hasUnderscore :: String -> Bool
+hasUnderscore = elem '_'
+
+hasCapitalism :: String -> Bool
+hasCapitalism s = any isAllUpper (trigrams s)
+  where
+    isAllUpper = all isUpper
+
+trigrams :: String -> [String]
+trigrams = \case
+  a:b:c:as -> [a,b,c] : trigrams (b:c:as)
+  _otherwise -> []
+
+--- these are copied from Hint.Naming ---
+
+shorten :: LHsDecl GhcPs -> LHsDecl GhcPs
```

Comment:
Please add some comments explaining what this does.

Reply:
done.